### DISCUSSION TITLE
Unable to execute HTTP request: io/netty/handler/codec/http2/Http2Connection$Listener at software.amazon.awssdk.utils.CompletableFutureUtils.errorAsCompletionException

### DISCUSSION THREAD

**Discussion by katharineyau1** [aws/aws-sdk-java-v2#3468] [Q&A]
I was trying to follow the S3 example from here:

https://github.com/awsdocs/aws-doc-sdk-examples/blob/main/javav2/example_code/s3/src/main/java/com/example/s3/async/GetObjectDataAsync.java

Here is my code: 

`package Processes.TestS3ObjectOps;
import java.util.*;
import java.io.*;

import java.io.IOException;
import software.amazon.awssdk.core.sync.RequestBody;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;
import software.amazon.awssdk.services.s3.model.S3Exception;
//import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.http.nio.netty.NettyNioAsyncHttpClient;
import software.amazon.awssdk.services.s3.S3AsyncClient;
import software.amazon.awssdk.services.s3.model.ListBucketsRequest;
import software.amazon.awssdk.services.s3.model.ListBucketsResponse;
import java.util.concurrent.CompletableFuture;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;
import software.amazon.awssdk.services.s3.model.PutObjectResponse;
import java.nio.file.Paths;
import software.amazon.awssdk.core.async.AsyncRequestBody;
import software.amazon.awssdk.services.s3.model.GetObjectRequest;
import software.amazon.awssdk.services.s3.model.GetObjectResponse;
import software.amazon.awssdk.services.s3.model.S3Exception;
import software.amazon.awssdk.core.async.AsyncResponseTransformer;

public class TestS3ObjectOps{
	public TestS3ObjectOps() {

	
	}
	 public static void putObjectAsync(S3AsyncClient client,String bucketName, String key, String path) {

			PutObjectRequest objectRequest = PutObjectRequest.builder()
					.bucket(bucketName)
					.key(key)
					.build();

			CompletableFuture<PutObjectResponse> future = client.putObject(objectRequest,
					AsyncRequestBody.fromFile(Paths.get(path))
			);
			future.whenComplete((resp, err) -> {
				try {
					if (resp != null) {
						System.out.println("Object uploaded. Details: " + resp);
					} else {
						// Handle error.
						err.printStackTrace();
					}
				} finally {
					// Only close the client when you are completely done with it.
					client.close();
				}
			});

			future.join();
		}

    public static void getObject (S3AsyncClient s3AsyncClient, String bucketName, String keyName, String path ) {

        try {
            GetObjectRequest objectRequest = GetObjectRequest
                    .builder()
                    .key(keyName)
                    .bucket(bucketName)
                    .build();

            CompletableFuture<GetObjectResponse> futureGet = s3AsyncClient.getObject(objectRequest,
                    AsyncResponseTransformer.toFile(Paths.get(path)));

            futureGet.whenComplete((resp, err) -> {
                try {
                    if (resp != null) {
                        System.out.println("Object downloaded. Details: " + resp);
                    } else {
                        err.printStackTrace();
                    }
                } finally {
                    // Only close the client when you are completely done with it.
                    s3AsyncClient.close();
                }
            });
            futureGet.join();

        } catch (S3Exception e) {
            System.err.println(e.awsErrorDetails().errorMessage());
            System.exit(1);
        }
    }
	public void invoke() throws Exception {

		AwsBasicCredentials awsCreds = AwsBasicCredentials.create(
		  "my-aws-key-id",
		  "my-aws-key-secret");

				String bucketName = "mybucketName";
				String keyName = "000015001d95c0850f6c3ba046e5535d";
				String path = "d:\\temp\\000015001d95c0850f6c3ba046e5535d";

		S3AsyncClient s3AsyncClient = S3AsyncClient.builder()
		.credentialsProvider(StaticCredentialsProvider.create(awsCreds))
			.region(Region.US_EAST_1)
			.httpClientBuilder(NettyNioAsyncHttpClient.builder()
			  .maxConcurrency(10)
			  .maxPendingConnectionAcquires(1000))
			.build();

				getObject (s3AsyncClient, bucketName, keyName, path);
				s3AsyncClient.close();
		}
}
`

but getting the following exception:

`
While executing [invoke] encountered [java.util.concurrent.CompletionException] : [software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: io/netty/handler/codec/http2/Http2Connection$Listener at software.amazon.awssdk.utils.CompletableFutureUtils.errorAsCompletionException(CompletableFutureUtils.java:62)]
	at com.tibco.plugin.java.JavaActivity.eval(Unknown Source)
	at com.tibco.pe.plugin.Activity.eval(Unknown Source)
	at com.tibco.pe.core.TaskImpl.eval(Unknown Source)
	at com.tibco.pe.core.Job.a(Unknown Source)
	at com.tibco.pe.core.Job.k(Unknown Source)
	at com.tibco.pe.core.JobDispatcher$JobCourier.a(Unknown Source)
	at com.tibco.pe.core.JobDispatcher$JobCourier.run(Unknown Source)
`
Can anyone advise what this error is about? and what can probably go wrong with the code? Thanks!

**Comment by debora-ito** [comment]
@katharineyau1 can you provide the full stacktrace with the error and which version of the SDK are you using?

**Comment by katharineyau1** [comment]
@debora-ito , thanks for your reply. It is SDK version 2. It was found to be classdefnotfound issue. After resolving that, I now get the HTTP connection timeout issue which I'm trying to resolve from my end. 

Caused by: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: connection timed out: myasxtestbucket.s3.ap-southeast-2.amazonaws.com/52.95.132.98:443
